src/Compilers/Core/Portable/Compilation/Compilation.cs

```diff
@@ -1169,7 +1169,9 @@ public INamedTypeSymbol CreateNativeIntegerTypeSymbol(bool signed)
             {
                 val = CommonGetTypeByMetadataName(fullyQualifiedMetadataName);
                 var result = _getTypeCache.TryAdd(fullyQualifiedMetadataName, val);
-                Debug.Assert(result || (_getTypeCache.TryGetValue(fullyQualifiedMetadataName, out var addedType) && ReferenceEquals(addedType, val)));
+                Debug.Assert(result
+                 || !_getTypeCache.TryGetValue(fullyQualifiedMetadataName, out var addedType) // Could fail if the type was already evicted from the cache
```

Comment:
>// Could fail if the type was already evicted from the cache [](http://example.com/codeflow?start=94&length=60)

Technically, I think it is possible that the type got evicted and then a different instance got added by other threads.

Reply:
Or if we add a type we should always get the same instance?

Reply:
> Or if we add a type we should always get the same instance?

I believe it should always be the same instance. This API effectively just walks through every assembly and looks for the member symbols from the assembly, and (in C#) gets the public version. Those should not be changing.